### ISSUE TITLE
Marshaling map[int64]int64 or []CustomType

### ISSUE THREAD

**Issue by mauroartizzu** [guregu/dynamo#184]
Hello, sorry but this is unclear to me.

Would it be possible to marshal custom types like these? At least the map, since the second struct is a bit complex to marshal but I don't really need it, so it's ignored

```
type Receipt struct {
	PartitionedTotalVat map[int64]int64 `json:"partitioned_total_vat" dynamo:",set"`
	Items               []ReceiptItem   `json:"items" dynamo:"-"`
}
```
I receive "cannot marshal type map[int64]int64 into a set" as an error. I think I must implement some marshaling receivers in my Receipt struct but the documentation is not so clear about that. Thanks a lot!

**Comment by guregu** [comment]
Hello, 

First issue is you're declaring it to be a set with `dynamo:",set"`. Sets are collections of unique numbers/strings/bytes. If you remove this it will become a map instead. You can read more about the data types here: https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.NamingRulesDataTypes.html#HowItWorks.DataTypes

The second issue is that DynamoDB wants maps to have strings as their keys, and you're using int64 as the key. You could change your map to use strings like `map[string]int64`.
This library supports using [TextMarshaler](https://pkg.go.dev/encoding#TextMarshaler) as keys as well. So you could implement a custom marshaler like this:

```go
type MyID int64

func (n MyID) MarshalText() ([]byte, error) {
	return []byte(strconv.FormatInt(int64(n), 10)), nil
}

func (n *MyID) UnmarshalText(text []byte) error {
	x, err := strconv.ParseInt(string(text), 10, 64)
	if err != nil {
		return err
	}
	*n = MyID(x)
	return nil
}
```

TL;DR: remove `dynamo:",set"`, add custom MarshalText/UnmarshalText to map key

```go
type Receipt struct {
	PartitionedTotalVat map[MyID]int64 `json:"partitioned_total_vat"`
	Items               []ReceiptItem   `json:"items" dynamo:"-"`
}
```

BTW, I don't know what your ReceiptItem class looks like, but you can include structs within items and it will work. Structs become maps as well.

**Comment by mauroartizzu** [comment]
Thanks! it works :D it was not so clear to me that omitting ,set would parse the structure as a map. It works flawlessly and converting the map to a [string]int64 is good, not a big issue even without custom marshaling. btw thanks
